Add initial signoff and gpg-signing support

OdinVex opened the pull request:
Adds (initial) support for automatic commit-signoff (ex: `Signed-off-by: Odin Vex <44311901+OdinVex@users.noreply.github.com>`) and gpg-signing support on the `Options->General` tab. Do note that not all git servers have gpg-signing enabled, at which point it will simply throw an error (from git, safely), and you'll need to disable gpg-signing.

Some development environments **require** signoff or even gpg-signing to be enabled.

Note: This commit lacks localization for user-viewed text in the Settings Dialog, specifically lines 113-122 and 140-143. I wasn't sure if the texts could be worded with more clarity or not. Submitting as-is, for now.

OdinVex commented:
...I forgot to use `cl-fmt.sh` again, my bad.

OdinVex commented:
Maybe we could update the UI to reflect `gpg-signed` commits/pushes/tags. A lock icon similar to GitHub with a status tooltip of 'Verified' might be the nicest way to go. I think the following areas could use an icon:

Left of the branch name in the tracking tree, same situation on the branch details area on the right.

exactly-one-kas commented:
I'm not sure how this works
libgit2 doesn't support automatic `format.signOff`

and pgp signing support is currently not implemented in libgit2 and there are no plans to support it there
The implementation for Gittyup is currently being done by @akselmo

Do you want these options to control behavior outside of Gittyup?

OdinVex commented:
Gittyup signs off on my commits with these options in my git config (inside Gittyup). I have other stuff in the config for which specific gpg key to use. I don't know anything about your backend as far as how it uses git.

Alex-Kent commented:
> Edit: I forgot that last year I modified Gittyup to remove code that worked with commits to use my installation's git instead, to specifically support signoff and GPG because libgit2 **does** support `git_commit_create_with_signature` but that is required to be implemented via the software. I implemented it using host git. It was a hackish work-around I forgot about. It works, but I don't think Gittyup wants to use local git installations...even though that would bring 'your installed version of git' support.

Could you add your modifications to the patch you submitted?  A configuration option could be added to along the lines of "Enable signing and signoff (requires `git` to be installed)"  Turning that option on could switch to your code instead of going through the library.   If and when `libgit2` adds native support the option and your work-around code could be removed.

OdinVex commented:
> > Edit: I forgot that last year I modified Gittyup to remove code that worked with commits to use my installation's git instead, to specifically support signoff and GPG because libgit2 **does** support `git_commit_create_with_signature` but that is required to be implemented via the software. I implemented it using host git. It was a hackish work-around I forgot about. It works, but I don't think Gittyup wants to use local git installations...even though that would bring 'your installed version of git' support.
> 
> Could you add your modifications to the patch you submitted? A configuration option could be added along the lines of "Enable signing and signoff (requires `git` to be installed)" Turning that option on could switch to your code instead of going through the library. If and when `libgit2` adds native support the option and your work-around code could be removed.

I thought about that at first but realized that my changes are also specific to my system, I even used hard-coded paths and a lot of hackish code. I think it currently unsuitable for public use. Not only that, but I'm not that big on Qt, so it's a bit labored.

As far as libgit2 is concerned, it *has* support for allowing the *client* to specify signed objects and stuff, it requires the client to handle the means of handing data off to be signed by the host (or whatever non-libgit2 means) and libgit2 can use it with `git_commit_create_with_signature`. For signoff, I believe Gittyup has something like that 'kind of' happening in the code, but I'm not sure. It's more like an invented work-around 'featurette' of Gittyup. Could be wrong on that bit, though, I'm exhausted lately.

Alex-Kent commented:
See my edit regarding GPG

OdinVex commented:
> See my edit regarding GPG

Just seen, and I agree, I'd rather let a git binary installed do all of the work, it'd be known standard and compatible without any possible hiccups, and that's why I went that way with my local src (several versions old but also contains all of the fixes, some stuff is just half-hazardly pulled in). I merely meant that libgit2 does include the ability to commit with a signature, but requires the client to handle the actual signing and such.

OdinVex commented:
A hackish work-around for now might be to have Gittyup automatically append to commit messages `Signed-off-by: Herp Derp <herp.derp@example.fqdn>` if it doesn't end with it already, but it's hackish..eh. Before I implemented them in my binary, I simply swapped to terminal and did a commit signed with signoff as well. `git commit -S -s -m "Herp Derp added" -m "Description of changes."` \*shrug\*

Alex-Kent commented:
Kernighan and Pike's words come to mind: [*Program design in the UNIX environment*](https://harmful.cat-v.org/cat-v/unix_prog_design.pdf)

Alex-Kent commented:
To elaborate on my earlier comment regarding GPG:

Calling an external `gpg` binary to handle signing is pretty much a requirement.  Even if signing could be done in-application many organizations exert tight control over private keys and flat-out disallow access to such material except through a small number of trusted applications (e.g. `gpg`).  Doing the crypto in-application would (at the very least) make the signing feature unusable in such environments.

OdinVex commented:
> To elaborate on my earlier comment regarding GPG:
> 
> Calling an external `gpg` binary to handle signing is pretty much a requirement. Even if signing could be done in-application many organizations exert tight control over private keys and flat-out disallow access to such material except through a small number of trusted applications (e.g. `gpg`). Doing the crypto in-application would (at the very least) make the signing feature unusable in such environments.

I agree, but they seem intent on using Flatpak, -_-.

OdinVex commented:
I've managed to refactor the latest version of Gittyup to the point that the decision for a wrapper around calling gpg-signing can be implemented by adjusting `src/git/Repository.cpp`, specifically `Commit Repository::commit(const Signature &author, const Signature &committer, const QString &message, const AnnotatedCommit &mergeHead)`.

I've edited the function to use a buffer instead, so that the path the routine takes can transparently implement signing or optionally not. It's a preparatory refactor for adding in the simple “if gpg, sign” one-liner from a later-implemented wrapper. I haven't pushed that section to my fork yet, I'm wondering what the lead developer wishes to do as far as methods of implementation. Can Flatpak integrate gpg-related stuff? Can Flatpak 'escape' the sandbox (or can holes be poked in with Flatseal) to allow access to host gpg?

OdinVex commented:
Woops, ignore that push, only the Repository.cpp stuff is relevant right now. Will wait on pushing to my fork until done.

Murmele commented:
@OdinVex can you mark the mr as draft as long as it is not readyß

mwerle commented:
Would it be feasibly to split this PR into two separate items?  One for the "Signed-off-by" and one for the GPG work?  It seems to me that there is no holdup for the former, while there are decisions to be made for the GPG work to be completed.
